### PR TITLE
Add Mono.cache(ttl, ttl, ttl) overload with Scheduler

### PR DESCRIPTION
**Pull request by mickeelm**
Fix for #2580.

I could not find any tests for the other overloaded cache-operators. Please point me in the right direction if you want tests for this.

### CODE REVIEW

**Comment by simonbasle** [comment]
Hey @mickeelm thanks for the submission ! The tests for that operator are in `MonoCacheTimeTest.java`, can you add a few simple tests?

**Comment by mickeelm** [comment]
I checked the test class this weekend and if I understand correctly, these tests are more about testing the core cache functionality etc.

So, I guess what's reasonable here is to add a test using a VTS (testing each of value, error, empty), to test this particular overloaded method (as that's the only thing that's been added).  If so, would you like me to add dedicated tests in the same flavour for the other cache()-methods? Some of the cache()-methods are _used_ in tests but they don't have these kind of dedicated tests where we make sure that the actual method does what it supposed to (right?).

**Comment by simonbasle** [comment]
I think it would be interesting to use VTS, indeed, through the (new) public API.

There would need to be a test that passing a `null` `Scheduler` is rejected, too.

The other tests don't go through the public API, but the functionality is tested in the underlying class (`MonoCacheTime` is instantiated directly), so I think it is fine.

**Comment by mickeelm** [comment]
I'm going to push a few tests this evening. One interesting find is that if I pass a `null` ttl, an NPE is only thrown if that `null` ttl matches the signal. Easier to explain with an example.

This test, with a value Mono, that has `ttlForValue` set, passes (i.e. an NPE is thrown):

```
    @Test
    public void nullValueTtlInSignalDependentCacheCausesNpe(){
        Duration longTtl = Duration.ofDays(1);
        StepVerifier.create(Mono.just("hello").cache(null, e -> longTtl, () -> longTtl, Schedulers.parallel()))
                .expectErrorMatches(t -> t instanceof NullPointerException)
                .verify();
    }
```

But, if I change that to an empty Mono, the test fails (i.e. an NPE is not thrown):

```
@Test
    public void nullValueTtlInSignalDependentCacheCausesNpe(){
        Duration longTtl = Duration.ofDays(1);
        StepVerifier.create(Mono.empty().cache(null, e -> longTtl, () -> longTtl, Schedulers.parallel()))
                .expectErrorMatches(t -> t instanceof NullPointerException)
                .verify();
    }
```

So, I understand why this happens (because it only checks for the ttl-function corresponding to a signal once that signal actually comes, if it ever does). I just wanted to bring it up to make sure that it's intended to work this way.

**Comment by simonbasle** [comment]
That's already an interesting byproduct of having the tests 😄
I'd say that, since the `Mono.cache(...)` API doesn't mention `@Nullable` on these `Function`s, it would be better to reject `null` early. In that case, the test wouldn't even attempt to subscribe to the `MonoCacheTime`, as the constructor would throw an NPE.

See how `map` behaves for instance: https://github.com/reactor/reactor-core/blob/b8a5b8132f72b8816f6e819d5198b175df18183e/reactor-core/src/main/java/reactor/core/publisher/FluxMap.java#L51

**Comment by mickeelm** [comment]
Glad to see early payback on the test investment :smile: 

I changed to your suggested approach regarding null arguments in the production code and added null-tests for all of the cache-operators. One thing I thought about doing is changing the variable name `clock` in `MonoCacheTime` to `timer` in order to match the API naming (the error message `"clock must not be null"` is quite confusing when the API name is `timer`). If there isn't a special reason to keep the variable name `clock`, just let me know and I'll push a change.

File: reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java [hunk 1]
```diff
@@ -900,4 +905,57 @@ public void scanSubscriber() {
...
 	}
 
+	@Test
+    public void signalDependentWithSchedulerOperatorTests() {
+	    Duration valueTtl = Duration.ofMillis(1000);
+        Duration errorTtl = Duration.ofMillis(2000);
+        Duration emptyTtl = Duration.ofMillis(3000);
+
+        AtomicInteger subCount = new AtomicInteger();
+        Supplier<Mono<Integer>> valueMonoSupplier = () -> Mono.just(subCount.incrementAndGet());
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java:915]: I think the whole test with the `Supplier`, `Function` using a `Mono.defer(monoSupplier)`, etc... can be simplified:
 - turn these `Supplier`s into `Mono`s, but do pass each lambda to a `Mono.defer(lambda)` call (ie the lazyness of incrementing `subCount` comes from Mono.defer, not a Supplier)
 - make the `monoFunction` a `Function<Mono<Integer>, Mono<Integer>>`
 - remove the deferring from the function now that it is taken care of above

File: reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java [hunk 2]
```diff
@@ -900,4 +905,57 @@ public void scanSubscriber() {
...
+
+        assertCacheTtl(monoFunction.apply(valueMonoSupplier), valueTtl, subCount, vts);
+        assertCacheTtl(monoFunction.apply(errorMonoSupplier), errorTtl, subCount, vts);
+        assertCacheTtl(monoFunction.apply(emptyMonoSupplier), emptyTtl, subCount, vts);
+    }
+
+    private void assertCacheTtl(Mono<Integer> mono, Duration ttl, AtomicInteger subCount, VirtualTimeScheduler vts) {
+	    subCount.set(0);
+
+        mono.subscribe();
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java:931]: this leads to some log pollution in the case of the `errorMonoSupplier`, because no error handler is defined. To avoid that, better use `mono.subscribe(v -> {}, e -> {});` at the 3 spots where you used `mono.subscribe();`.

**Review by simonbasle — CHANGES_REQUESTED** [review]
yeah the formatting should be cleaned up 😉 

as for renaming `clock` to `timer` in constructor, I don't have any strong opinion...

I just think it is hard to guarantee that the public API parameter names will always keep in sync with the operator's constructor parameter names (or field names for that matter). I think it is fine, though. Even if the names are slightly different they should still be disambiguated enough that mentally mapping the two sets of names shouldn't be a big issue 😄 (and the stacktrace will still be helpful).

File: reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java (hunk 1 above)
Comment (mickeelm) [reply reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java:915]: Of course, that's a lot cleaner. Willdo!

File: reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java (hunk 2 above)
Comment (mickeelm) [reply reactor-core/src/test/java/reactor/core/publisher/MonoCacheTimeTest.java:931]: Perfect,  I'll fix it

**Comment by mickeelm** [comment]
- I'll let the naming stay as is (and yeah, there's no guarantee that the names will be in sync, that's a good point)
- I'll fix the tests per your comments
- I'll fix the formatting, of course

I'll push as soon as it's done.

**Comment by simonbasle** [comment]
@mickeelm we have exported eclipse and intellij code formatting configurations in `/codequality` folder, but that might be a bit outdated. yeah we still have a bunch of files where there is a mix of spaces and tabs even outside of fluent method calls `.` alignment, which is confusing 😞

**Comment by mickeelm** [comment]
> @mickeelm we have exported eclipse and intellij code formatting configurations in `/codequality` folder, but that might be a bit outdated. yeah we still have a bunch of files where there is a mix of spaces and tabs even outside of fluent method calls `.` alignment, which is confusing 😞

Aaah didn't know, but I'll try that the next time. This time I used vim with tab characters 🔍 to make sure it was right. So I hope it is with the latest commit. But for the next time I'll definitely try with the settings file.

**Comment by simonbasle** [comment]
yeah reducing the friction around formatting is on my mind. having a `.editorconfig` might help at least with tab indentation, but only intellij seems to have the ability to "reflow" fluent method calls by adding spaces for alignment of each step. But that's a separate discussion to have 😅

**Review by simonbasle — APPROVED** [review]
